### PR TITLE
Move DocumentFilterManager into cluster DI container

### PR DESCRIPTION
**Pull request by brantburnett**
Motivation
----------
Make the DocumentFilterManager scoped to the cluster instead of a
static singleton. This will better support multiple clusters which may
have different requirements, as well as unit testing injection.

Modifications
-------------
Rewrite DocumentFilterManager to be an instance class instead of a
singleton. Perform some cleanup as well, such as making
DocumentFilterSet truly immutable, using nullable ref types, and fixing
up the XML doc comments.

Create extensions for ClusterOptions that register Linq2Couchbase on the
cluster services and allow extensibility.

Update some of the documentation with a modern example

Results
-------
DocumentFilterManager now lives on the ICluster.ClusterServices for
each cluster, with independent configuration.

To use Linq2Couchbase, AddLinq *must* be called on ClusterOptions prior
to connecting to the cluster.

Relates to #324

### CODE REVIEW

**Comment by brantburnett** [comment]
@mgroves I'd also like your opinion on this PR, in terms of client usability. I think it's a lot more extensible and functional long-term, but it does carry the downside that AddLinq must be called on ClusterOptions when connecting to the cluster. I think it's worth it, given how it meshes with the .NET Core style, but welcome second opinions. See the updated README.md in the PR for an example.

File: README.md [hunk 1]
```diff
@@ -1,69 +1,75 @@
...
+        options.Password = "password";
 
-    var context = new BucketContext(ClusterHelper.GetBucket("travel-sample"));
+        options.AddLinq();
+    });
+
+    // Wait until the cluster is bootstrapped
+    await cluster.WaitUntilReadyAsync(TimeSpan.FromSeconds(10));
+
+    var context = new BucketContext(await cluster.BucketAsync("travel-sample"));
```
Comment (jeffrymorris) [inline README.md:48]: Should Cluster have an extension method that creates BucketContexts? 

`cluster.GetBucketContext(string name)` or `cluster.GetQueryContext(string bucketName)`?

Open to counter-ideas here.

Comment (brantburnett) [reply README.md:48]: That could be interesting. I'd probably go with CreateBucketContext to make clear it's making a new one. I'd also think an extension on IBucketProvider from the DI extensions would be good, too. Feels like a separate PR, though.

Comment (jeffrymorris) [reply README.md:48]: @brantburnett - this ready for review?

Comment (brantburnett) [reply README.md:48]: I think so, up to you if you want me to add more to it now or in a separate PR.

**Pull request merged**